### PR TITLE
FedCurv algorithm

### LINKED ISSUE #66: Add FedCurv Algorithm
https://arxiv.org/pdf/1910.07796.pdf

We'd like to support FedCurv for robust aggregation.

### PR DESCRIPTION
**Pull request by itrushkin**
Implements "Federated Curvature" algorithm from paper: [Overcoming Forgetting in Federated Learning on
Non-IID Data](https://arxiv.org/pdf/1910.07796.pdf).
Closes #66.

### CODE REVIEW

**Comment by itrushkin** [comment]
[TensorBoard](https://tensorboard.dev/experiment/BEW8JXRQQKOROLR9zXNZgQ/)

**Comment by itrushkin** [comment]
Jenkins please retry a build

File: openfl-tutorials/interactive_api/PyTorch_FedCurv/envoy/tinyimagenet_shard_descriptor.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+class TinyImageNetDataset(ShardDataset):
+    """TinyImageNet shard dataset class."""
+
+    NUM_IMAGES_PER_CLASS = 500
+
+    def __init__(self, data_folder: Path, data_type='train', rank=1, worldsize=1):
+        """Initialize TinyImageNetDataset."""
+        self.data_type = data_type
+        self._common_data_folder = data_folder
+        self._data_folder = os.path.join(data_folder, data_type)
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_FedCurv/envoy/tinyimagenet_shard_descriptor.py]:
```suggestion
        self._data_folder = data_folder / data_type
```

File: openfl-tutorials/interactive_api/PyTorch_FedCurv/envoy/tinyimagenet_shard_descriptor.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        self.labels = {}  # fname - label number mapping
+        self.image_paths = sorted(
+            glob.iglob(
+                os.path.join(self._data_folder, '**', '*.JPEG'),
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_FedCurv/envoy/tinyimagenet_shard_descriptor.py]:
```suggestion
                str(self._data_folder / '**' / '*.JPEG'),
```

File: openfl-tutorials/interactive_api/PyTorch_FedCurv/envoy/tinyimagenet_shard_descriptor.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+                recursive=True
+            )
+        )[rank - 1::worldsize]
+        with open(os.path.join(self._common_data_folder, 'wnids.txt'), 'r') as fp:
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_FedCurv/envoy/tinyimagenet_shard_descriptor.py]:
```suggestion
        with open(self._common_data_folder / 'wnids.txt', 'r') as fp:
```

File: openfl/utilities/fedcurv/torch/fedcurv.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            target = target.to(device)
+            output = model(sample)
+            loss = F.nll_loss(F.log_softmax(output, dim=1), target)
+            loss.backward()
+
+            for n, p in model.named_parameters():
+                if p.requires_grad:
+                    precision_matrices[n].data += p.grad.data ** 2 / len(data_loader)
+
+        precision_matrices = {n: p for n, p in precision_matrices.items()}
```
Comment (alexey-khorkin) [inline openfl/utilities/fedcurv/torch/fedcurv.py]: Is this line of code really necessary?

File: openfl/utilities/fedcurv/aggregation_function.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    def call(self, local_tensors, db_iterator, tensor_name, fl_round, tags):
+        """Apply aggregation."""
+        if (
+            tensor_name.endswith('_u')
+            or tensor_name.endswith('_v')
+        ):
+            tensors = [local_tensor.tensor for local_tensor in local_tensors]
+            agg_result = np.sum(tensors, axis=0)
+            with open(Path('~').expanduser() / 'fedcurv_log.txt', 'a') as f:
```
Comment (dmitryagapov) [inline openfl/utilities/fedcurv/aggregation_function.py]: If `Path('~').expanduser()` user's home direcory maybe better to use `Path.home()`

**Review by dmitryagapov — CHANGES_REQUESTED** [review]

File: openfl/utilities/fedcurv/torch/fedcurv.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                u_global, v_global = (
+                    get_buffer(model, target).detach()
+                    for target in (f'{name}_u', f'{name}_v')
+                )
+                u_local, v_local = (
+                    getattr(self, name).detach()
+                    for name in (f'{name}_u', f'{name}_v')
+                )
+                if name == 'fc2.bias':
+                    with open(Path('~').expanduser() / 'fedcurv_log.txt', 'a') as f:
```
Comment (dmitryagapov) [inline openfl/utilities/fedcurv/torch/fedcurv.py]: `Path('~').expanduser()` -> `Path.home()`

**Review by dmitryagapov — CHANGES_REQUESTED** [review]

File: openfl/utilities/fedcurv/torch/__init__.py [hunk 7]
```diff
@@ -0,0 +1,2 @@
+"""openfl.utilities.fedcurv.torch package."""
```
Comment (alexey-khorkin) [inline openfl/utilities/fedcurv/torch/__init__.py:3]: Should we add corporation comments?
```suggestion
# Copyright (C) 2021-2022 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
"""openfl.utilities.fedcurv.torch package."""
```

File: openfl/utilities/fedcurv/aggregation_function.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
+"""FedCurv Aggregation function module."""
```
Comment (alexey-khorkin) [inline openfl/utilities/fedcurv/aggregation_function.py]: Should we replace this module to https://github.com/intel/openfl/tree/develop/openfl/component/aggregation_functions and rename it to `fedcurv_aggregation.py` ?

**Comment by itrushkin** [comment]
Experiments results:
image
Graphs show that FedCurv with lambda equal to 10 performs better than FedAvg in terms of accuracy, and some other lambda values (1000, 10000000).

**Comment by alexey-khorkin** [comment]
> Experiments results: image Graphs show that FedCurv with lambda equal to 10 performs better than FedAvg in terms of accuracy, and some other lambda values (1000, 10000000).
@itrushkin 
Is `histology_test_experiment` the FedAvg setup?

**Comment by itrushkin** [comment]
> > Experiments results: image Graphs show that FedCurv with lambda equal to 10 performs better than FedAvg in terms of accuracy, and some other lambda values (1000, 10000000).
> > @itrushkin
> > Is `histology_test_experiment` the FedAvg setup?

@alexey-khorkin Yes, correct. Perhaps the difference could be much more visible with number of rounds increasing.

File: openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+class HistologyShardDescriptor(ShardDescriptor):
+    """Shard descriptor class."""
+
+    URL = ('https://zenodo.org/record/53169/files/Kather_'
+           'texture_2016_image_tiles_5000.zip?download=1')
+    FILENAME = 'Kather_texture_2016_image_tiles_5000.zip'
+    ZIP_SHA384 = ('7d86abe1d04e68b77c055820c2a4c582a1d25d2983e38ab724e'
+                  'ac75affce8b7cb2cbf5ba68848dcfd9d84005d87d6790')
+    DEFAULT_PATH = Path.home() / '.openfl' / 'data'
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py]: Try to avoid using folders outside the current workspace. The default data location for envoy is ./data

File: openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    TRAIN_SPLIT_RATIO = 0.8
+
+    def __init__(self, data_folder: Path, data_type='train', rank=1, worldsize=1):
+        """Histology shard dataset class."""
+        self.data_type = data_type
+        root = Path(data_folder) / 'Kather_texture_2016_image_tiles_5000'
+        classes = [d.name for d in root.iterdir() if d.is_dir()]
+        class_to_idx = {cls_name: i for i, cls_name in enumerate(classes)}
+        self.samples = []
+        root = root.expanduser()
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py]:
```suggestion
        root = root.absolute()
```

File: openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py (hunk 9 above)
Comment (igor-davidyuk) [reply openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py]:
```suggestion
    DEFAULT_PATH = Path('.') / 'data'
```

File: openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        for target_class in sorted(class_to_idx.keys()):
+            class_index = class_to_idx[target_class]
+            target_dir = os.path.join(root, target_class)
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py]: You are already using pathlib, os.path is redundant

Comment (igor-davidyuk) [reply openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py]:
```suggestion
            target_dir = root / target_class
```

File: openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+            for class_root, _, fnames in sorted(os.walk(target_dir, followlinks=True)):
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/histology_shard_descriptor.py]: may be we can use Path.glob() here

File: openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/populate_envoys.sh [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+for i in {1..8}
+do
+    mkdir $i
+    echo "shard_descriptor:
+    template: histology_shard_descriptor.HistologyShardDescriptor
+    params:
+        data_folder: histology_data
+        rank_worldsize: $i,8
+" > $i/envoy_config.yaml
```
Comment (igor-davidyuk) [inline openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/populate_envoys.sh]: well done 👍

**Review by igor-davidyuk — APPROVED** [review]
Please make sure data is downloaded and used correctly with a new path

**Review by dmitryagapov — APPROVED** [review]

**Review by aleksandr-mokrov — APPROVED** [review]

File: openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/populate_envoys.sh [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+    echo "fx envoy start -n env_$i --disable-tls --envoy-config-path envoy_config.yaml -dh nnlicv431.inn.intel.com -dp 50053" > $i/start_envoy.sh
```
Comment (alexey-gruzdev) [inline openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/envoy/populate_envoys.sh]: Can we parametrize `-dh nnlicv431.inn.intel.com -dp 50053` with variables, and maybe insert everywhere `localhost`?

**Comment by itrushkin** [comment]
@igor-davidyuk Re-launched experiment and data is being split and loaded properly.

File: openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/README.md [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+cd envoy
+bash populate_envoys.sh # This creates all envoys folders in current directory
+pip install -r requirements.txt
+bash start_envoys.sh # This launches all envoy instances 
+```
+
+3. In a separate terminal, launch a Jupyter Notebook:
+```
+cd workspace
+jupyter notebook
```
Comment (alexey-gruzdev) [inline openfl-tutorials/interactive_api/PyTorch_Histology_FedCurv/README.md]: `jupyter lab pytorch_histology.ipynb`?

**Review by alexey-gruzdev — APPROVED** [review]
